File: _CodeTopics/LeetCode/201-400/000210/000210.py
class Solution(object):
    
    hasRing = False

    def findOrder(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: List[int]
        """

        # 和 `LC207. 课程表` 官方答案很像，下面这个提交也是改了下官方答案，更易懂。
        # https://leetcode-cn.com/submissions/detail/94672410/

        visited = [False for i in range(numCourses)]
        res = []
        
        edges = collections.defaultdict(list)
        for info in prerequisites:
            edges[info[1]].append(info[0])
        
        def dfs(vertex):
            visited[vertex] = "ongoing"
            for v in edges[vertex]:
                if visited[v] == False:
                    dfs(v)
                    if self.hasRing:
                        return
                elif visited[v] == "ongoing":
                    self.hasRing = True
                    return
            visited[vertex] = True
            # LC207 的话就没必要搞这个栈了，因为只需要判断拓扑排序是否存在。
            # 而这道题要求如果存在的话要输出一个拓扑排序。
            res.append(vertex)

        for i in range(numCourses):
            if not self.hasRing and not visited[i]:
                dfs(i)
        return res[::-1] if not self.hasRing else []
        
"""
https://leetcode-cn.com/submissions/detail/248245338/

执行用时：20 ms, 在所有 Python 提交中击败了94.76%的用户
内存消耗：15.8 MB, 在所有 Python 提交中击败了10.49%的用户
通过测试用例：
44 / 44
"""

File: _CodeTopics/LeetCode/201-400/000210/tran/000210.cpp
// https://leetcode-cn.com/submissions/detail/248245338/

class Solution {
public:
    bool hasRing = false;
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> visited(numCourses, 0);  // 0 表示未访问过的节点
        unordered_map<int, vector<int>> edges;
        for (vector<int> info : prerequisites) {
            edges[info[1]].push_back(info[0]);
        }
        
        vector<int> res;
        for (int i = 0; i < numCourses; i++) {
            if (!hasRing && visited[i] == 0) {
                dfs(i, visited, edges, res);
            }
        }
        
        if (hasRing) {
            return {};
        }
        reverse(res.begin(), res.end());
        return res;
    }
private:
    void dfs(int vertex, vector<int>& visited, unordered_map<int, vector<int>>& edges, vector<int>& res) {
        visited[vertex] = 1;  // 1 表示正在访问的节点
        for (int v : edges[vertex]) {
            if (visited[v] == 0) {
                dfs(v, visited, edges, res);
                if (hasRing) {
                    return;
                }
            }
            else if (visited[v] == 1) {
                hasRing = true;
                return;
            }
        }
        visited[vertex] = 2;  // 2 表示已经访问结束的节点
        res.push_back(vertex);
    }
};

/*
https://leetcode.cn/problems/course-schedule-ii/submissions/464948361/

时间
详情
24ms
击败 39.64%使用 C++ 的用户
内存
详情
14.74MB
击败 6.59%使用 C++ 的用户
*/
